# Inference of ancestral linkage groups

## 1. Description
Macrosynteny between annotated chromosome-level genomes of 11 annelid species plus two outgroups (a nemertean - *Lineus longissimus* - and a mollusk - *Pecten maximus*) was explored with odp v0.3.0. The Bilaterian-Cnidarian-Sponge Linkage Groups (BCnS LGs) were inferred to describe the relation between these linkage groups and the chromosomes of the species in our dataset. For clarity, we merged the linkage groups that were always together in the outgroups and the marine annelids as follows: A1 comprises A1a and A1b, E comprises Ea and Eb, H_Q comprises H and Q, J2_L comprises J2 and L, K_O2 comprises K and O2, O1_R comprises O1 and R and Q comprises Qa, Qb, Qc and Qd. We referred to these linkage groups as merged linkage groups (mergedLGs). Additionally, we used odp v0.3.071 to infer linkage groups specific for the leeches (named LeechLGs) using the genomes of the three leeches, and for Crassiclitellata (ie, earthworms, named CrassiLGs) using the genomes of one earthworm per family represented (*Norana najaformis* representing Hormogastridae, *Eisenia andrei* representing Lumbricidae and *Metaphire vulgaris* representing Megascolecidae).  Given that *M. vulgaris* experienced a recent genome duplication, pairs of linkage groups corresponding to homologous chromosomes and the same chromosomes in the other two earthworms, were merged. To infer linkage groups for Clitellata (ClitLGs), the LeechLGs and the CrassiLGs were manually merged. To test for the enrichment of a specific linkage group in the set of chromosomes of a species, a Fisher test was used. The list of p-values was corrected using the Benjamini-Hochberg correction method. The ribbon plots and the oxford dotplots were generated using custom R scripts using the ggplot2 and RIdeogram packages and the output of odp. To identify the genomic regions that corresponded to each linkage group in each species, a custom script was used. Briefly, the region between a pair of genes corresponding to the same linkage group was considered to belong to that linkage group as long as there was not more than one gene belonging to a different linkage group.

![Figure_Synteny](Figure_Synteny.png)

## 2. Files in this repository
  - **blocksFromodp.R**: script to obtain the coordinates in each chromosome that correspond to a given linkage group.
  - **getPlots.R**: script to obtain the ribbon plots and the oxford dotplots.
